rust/otap-dataflow/crates/engine/src/exporter.rs

```diff
@@ -87,106 +266,175 @@ impl<PData> MessageChannel<PData> {
 
-impl<Msg> Clone for EffectHandler<Msg> {
-    fn clone(&self) -> Self {
-        EffectHandler {
-            exporter_name: self.exporter_name.clone(),
-            pd: self.pd,
-        }
-    }
-}
+                    // 1) Any pdata?
+                    pdata = self.pdata_rx.recv() => match pdata {
+                        Ok(pdata) => return Ok(Message::PData(pdata)),
+                        Err(_) => {
+                            // pdata channel closed → emit Shutdown
+                            let shutdown = self.pending_shutdown
+                                .take()
+                                .expect("pending_shutdown must exist");
+                            self.shutting_down_deadline = None;
+                            self.shutting_down = true;
+                            return Ok(Message::Control(shutdown));
+                        }
+                    },
 
-impl<Msg> EffectHandler<Msg> {
-    /// Creates a new `EffectHandler` with the given exporter name.
-    pub fn new<S: AsRef<str>>(exporter_name: S) -> Self {
-        EffectHandler {
-            exporter_name: Rc::from(exporter_name.as_ref()),
-            pd: std::marker::PhantomData,
+                    // 2) Deadline hit?
+                    _ = sleep_until_deadline.as_mut().expect("sleep_until_deadline must exist") => {
+                        let shutdown = self.pending_shutdown
+                            .take()
+                            .expect("pending_shutdown must exist");
+                        self.shutting_down_deadline = None;
+                        return Ok(Message::Control(shutdown));
+                    }
+                }
+            }
+
+            // Normal mode: no shutdown yet
+            tokio::select! {
+                biased;
+
+                // A) Control first
+                ctrl = self.control_rx.recv() => match ctrl {
+                    Ok(ControlMsg::Shutdown { deadline, reason }) => {
+                        if deadline.is_zero() {
+                            // Immediate shutdown, no draining
+                            self.shutting_down = true;
+                            return Ok(Message::Control(ControlMsg::Shutdown { deadline: Duration::ZERO, reason }));
+                        }
+                        // Begin draining mode, but don’t return Shutdown yet
+                        let when = Instant::now() + deadline;
+                        self.shutting_down_deadline = Some(when);
+                        self.pending_shutdown = Some(ControlMsg::Shutdown { deadline: Duration::ZERO, reason });
+                        continue; // re-enter the loop into draining mode
+                    }
+                    Ok(msg) => return Ok(Message::Control(msg)),
+                    Err(e)  => return Err(e),
+                },
+
+                // B) Then pdata
+                pdata = self.pdata_rx.recv() => {
+                    return pdata.map(Message::PData);
```

Comment:
Do we need to handle errors here? If the sender closes the channel from their end, maybe we could trigger shutdown then as well?

Reply:
Excellent point.

Reply:
Resolved